network/p2p/inspector/internal/cache/cache.go

```diff
@@ -0,0 +1,255 @@
+	backData := herocache.NewCache(config.sizeLimit,
+		herocache.DefaultOversizeFactor,
+		// this cache is supposed to keep the cluster prefix topics received record for the authorized (staked) nodes. Since the number of such nodes is
+		// expected to be small, we do not eject any records from the cache. The cache size must be large enough to hold all
+		// the records of the authorized nodes. Also, this cache is keeping at most one record per peer id, so the
+		// size of the cache must be at least the number of authorized nodes.
+		heropool.NoEjection,
+		config.logger.With().Str("mempool", "gossipsub=cluster-prefix-topics-received-records").Logger(),
+		config.collector)
+	recordCache := &RecordCache{
+		recordEntityFactory: recordEntityFactory,
+		decayFunc:           defaultDecayFunction(config.recordDecay),
+		c:                   stdmap.NewBackend(stdmap.WithBackData(backData)),
+	}
+	b := make([]byte, 100)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, err
+	}
+	recordCache.activeClusterIdsCacheId = flow.HashToID(b)
+	recordCache.initActiveClusterIds()
+	return recordCache, nil
+}
+
+// Init initializes the record cache for the given peer id if it does not exist.
+// Returns true if the record is initialized, false otherwise (i.e.: the record already exists).
+// Args:
+// - originId: the origin id the sender of the control message.
+// Returns:
+// - true if the record is initialized, false otherwise (i.e.: the record already exists).
+// Note that if Init is called multiple times for the same peer id, the record is initialized only once, and the
+// subsequent calls return false and do not change the record (i.e.: the record is not re-initialized).
+func (r *RecordCache) Init(originId flow.Identifier) bool {
+	entity := r.recordEntityFactory(originId)
+	return r.c.Add(entity)
+}
+
+// Update applies an adjustment that increments the number of cluster prefixed topics received by a peer.
+// Returns number of cluster prefix topics received after the adjustment. The record is initialized before
+// the adjustment func is applied that will increment the Counter.
+// It returns an error if the adjustFunc returns an error or if the record does not exist.
+// Assuming that adjust is always called when the record exists, the error is irrecoverable and indicates a bug.
+// Args:
+// - originId: the origin id the sender of the control message.
+// - adjustFunc: the function that adjusts the record.
+// Returns:
+//   - The number of cluster prefix topics received after the adjustment.
+//   - error if the adjustFunc returns an error or if the record does not exist (ErrRecordNotFound).
+//     All errors should be treated as an irrecoverable error and indicates a bug.
+//
+// Note if Adjust is called under the assumption that the record exists, the ErrRecordNotFound should be treated
+// as an irrecoverable error and indicates a bug.
+func (r *RecordCache) Update(originId flow.Identifier) (float64, error) {
```

Comment:
⚠️ A general note for all methods in this cache that accept a `flow.Identifier`:

The purpose of this cache is to maintain a mapping between `peer.ID`s and `RecordEntity`s, which should be transparent to any external module using it. However, the current implementation expects some translation between `peer.ID` and `flow.Identifier` to be performed externally. For instance, the inspector takes the hash of `peer.ID` when calling the `Inc` method of the active cluster ID tracker, which is connected to this cache's `Update` method.

The required translation exists because this cache internally uses `HeroCache`, which operates based on the concept of `flow.Identifier` (key) and `flow.Entity` (value). Thus, it is necessary to convert `peer.ID` to `flow.Identifier` to make it compatible with the underlying cache. However, this conversion should be done internally by the cache rather than externally by remote callers like the inspector, which do not directly interact with the cache. This approach violates the encapsulation principle of OOP. Moreover, having `flow.Identifier` in the method signatures of this cache is misleading, as it could be interpreted as the identifier of the node rather than the hash of its `peer.ID`. Finally, the conversion between `peer.ID` and `flow.Identifier` is commonly done through the identity provider in other parts of the code, and this translation by taking the hash of `peer.ID` is inconsistent with other code sections and inefficient. Given these concerns, I suggest the following:

1. Instead of externally taking the hash of `peer.ID` to convert it to `flow.Identifier`, please use an identity provider in the `ControlMsgValidationInspector`. Before making any call to this cache's methods, use the `ByPeerID` method of the `IdentityProvider` to translate the `peer.ID` to `flow.Identifier`:
  https://github.com/onflow/flow-go/blob/796f801427998ae53cdf847fe0ea30dfac322fe8/module/id_provider.go#L33-L39

2. If `ByPeerID` returns `false`, it should be considered an RPC validation error for cluster IDs. This indicates that an unstaked (or public) node is sending a message on a private cluster ID channel, which should not be allowed.

In this manner, the `flow.Identifier` type in the signature of the cache methods accurately represents its intended purpose, which is to signify a staked identity. This approach also appears to be more efficient, as it eliminates the need for an additional hash computation per cache query.

Reply:
https://github.com/onflow/flow-go/pull/4189/commits/1cfda0ce7020274d4c174204dc8e9e9109c3618e